api.planx.uk/webhooks/sanitiseApplicationData/operations.ts

```diff
@@ -0,0 +1,214 @@
+import { gql } from "graphql-request";
+import { subMonths } from "date-fns";
+
+import { Operation, OperationResult } from "./types";
+import { adminGraphQLClient } from "../../hasura";
+import { runSQL } from "../../hasura/schema";
+
+const RETENTION_PERIOD_MONTHS = 6;
+export const getRetentionPeriod = () => subMonths(new Date(), RETENTION_PERIOD_MONTHS);
+
+/**
+ * List of data sanitation operations
+ * XXX: Analytics logs do not contain application data
+ */
+ export const getOperations = (): Operation[] => ([
+  // Raw application data
+  sanitiseLowcalSessions,
+
+  // Audit records
+  sanitiseSessionBackups,
+  sanitiseUniformApplications,
+  sanitiseBOPSApplications,
+  sanitiseReconciliationRequests,
+
+  // Event logs
+  deleteHasuraEventLogs,
+]);
+
+export const operationHandler = async (operation: Operation): Promise<OperationResult> => {
+  let operationResult: OperationResult = {
+    operationName: operation.name,
+    status: "processing"
+  };
+
+  try {
+    const result = await operation()
+    operationResult = {
+      ...operationResult,
+      status: "success",
+      count: result.length,
+    };
+  } catch (error) {
+    operationResult = {
+      ...operationResult,
+      status: "failure",
+      errorMessage: (error as Error).message,
+    }
+  };
+
+  return operationResult;
+}
+
+export const sanitiseLowcalSessions: Operation = async () => {
+  const mutation = gql`
+    mutation SanitiseLowcalSessions($retentionPeriod: timestamptz) {
+      update_lowcal_sessions(
+        _set: { 
+          data: {}
+          email: ""
+          sanitised_at: "now()"
+        }
+        where: {
+          sanitised_at: { _is_null: true }
+          _or: [
+            { deleted_at: { _lt: $retentionPeriod } }
+            { submitted_at: { _lt: $retentionPeriod } }
+          ]
+        }
+      ) {
+        returning {
+          id
+        }
+      } 
+    }
+  `;
+  const { update_lowcal_sessions: {
+    returning: result
+  } } = await adminGraphQLClient.request(
+    mutation,
+    { retentionPeriod: getRetentionPeriod() },
+  );
+  return result;
+};
+
+export const sanitiseSessionBackups: Operation = async () => {
+  const mutation = gql`
+    mutation SanitiseSessionBackups($retentionPeriod: timestamptz) {
+      update_session_backups(
+        _set: {
+          flow_data: null
```

Comment:
nit: I wonder if rather than `null` or `{}` we could do populate flow_data with something like `_SANITISED_` here? Similar to how you can find `_REMOVED_` in `users` table currently? But I recognize that may make less sense/not be possible with a JSONB data type the same as a text field.

Reply:
noticed a new detail here today related to this: replacing users.email with `_REMOVED_` isn't actually sustainable because there's a unique constraint :see_no_evil: so scratch my `_SANITISED_` suggestion here and let's just aim for consistent soft delete processes / date fields :white_check_mark:

Reply:
Yeah this was my first idea and I hit this exact issue with emails also! 

I also considered something `{ "data": "**REDACTED**" }` for all JSONB column - I figured it might be visually helpful but in reality being able to easily query join to `flows` and then checking `sanitised_at IS NOT NULL`  / `sanitised_at IS NULL` is what I think would prove most helpful.